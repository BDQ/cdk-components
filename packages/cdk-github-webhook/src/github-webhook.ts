import * as path from 'path';
import { Construct, Duration, CustomResource } from '@aws-cdk/core';
import { SingletonFunction, Runtime, Code } from '@aws-cdk/aws-lambda';

export interface GithubWebhookProps {
    githubApiToken: string;
    githubRepoUrl: string;
    payloadUrl: string;
    events: string[]; // @see https://developer.github.com/v3/activity/events/types/
    logLevel?: 'debug' | 'info' | 'warning' | 'error';
}

export class GithubWebhook extends Construct {
    public constructor(
        parent: Construct,
        id: string,
        props: GithubWebhookProps,
    ) {
        super(parent, id);

        const handler = new SingletonFunction(this, 'CustomResourceHandler', {
            uuid: '83CBF3EB-7B62-44F2-8C67-8441E4C1232E',
            runtime: Runtime.NODEJS_10_X,
            code: Code.fromAsset(
                path.join(__dirname, '..', 'lambda', 'bundle.zip'),
            ),
            handler: 'lib/github-webhook.handler',
            lambdaPurpose: 'Custom::GithubWebhook',
            timeout: Duration.minutes(15),
        });

        const {
            githubApiToken,
            githubRepoUrl,
            payloadUrl,
            events,
            logLevel,
        } = props;

        new CustomResource(this, 'CustomResource', {
            serviceToken: handler.functionArn,
            resourceType: 'Custom::GithubWebhook',
            properties: {
                GithubApiToken: githubApiToken,
                GithubRepoUrl: githubRepoUrl,
                PayloadUrl: payloadUrl,
                Events: events,
                LogLevel: logLevel,
            },
        });
    }
}
