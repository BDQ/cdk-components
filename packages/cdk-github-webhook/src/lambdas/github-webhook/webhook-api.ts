import { Octokit } from '@octokit/rest';
import * as parseGithubUrl from 'parse-github-url';

export const createWebhook = async (
  githubApiToken: string,
  githubRepoUrl: string,
  payloadUrl: string,
  events: string[],
): Promise<Octokit.Response<Octokit.ReposCreateHookResponse>> => {
  const octokit = new Octokit({
    auth: githubApiToken,
  });

  const gh = parseGithubUrl(githubRepoUrl);

  if (gh === null || gh.owner === null || gh.name === null) {
    throw new Error('GithubRepoUrl is not correct');
  }

  const params = {
    name: 'web',
    owner: gh.owner,
    repo: gh.name,
    config: { url: payloadUrl, content_type: 'json' },
    events,
    active: true,
  };

  return octokit.repos.createHook(params);
};

export const updateWebhook = async (
  githubApiToken: string,
  githubRepoUrl: string,
  payloadUrl: string,
  events: string[],
  hookId: number,
): Promise<Octokit.Response<Octokit.ReposUpdateHookResponse>> => {
  const octokit = new Octokit({
    auth: githubApiToken,
  });

  const gh = parseGithubUrl(githubRepoUrl);

  if (gh === null || gh.owner === null || gh.name === null) {
    throw new Error('GithubRepoUrl is not correct');
  }

  const params = {
    owner: gh.owner,
    repo: gh.name,
    config: { url: payloadUrl, content_type: 'json' },
    events,
    active: true,
    hook_id: hookId,
  };

  return octokit.repos.updateHook(params);
};

export const deleteWebhook = async (
  githubApiToken: string,
  githubRepoUrl: string,
  hookId: number,
): Promise<Octokit.Response<Octokit.ReposDeleteResponse>> => {
  const octokit = new Octokit({
    auth: githubApiToken,
  });

  const gh = parseGithubUrl(githubRepoUrl);

  if (gh === null || gh.owner === null || gh.name === null) {
    throw new Error('GithubRepoUrl is not correct');
  }
  const params = {
    owner: gh.owner,
    repo: gh.name,
    hook_id: hookId,
  };

  return octokit.repos.deleteHook(params);
};
