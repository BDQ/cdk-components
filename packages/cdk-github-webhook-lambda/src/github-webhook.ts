import { CloudFormationCustomResourceEvent } from 'aws-lambda'; // eslint-disable-line import/no-unresolved

import {
    customResourceHelper,
    OnCreateHandler,
    OnUpdateHandler,
    OnDeleteHandler,
    ResourceHandler,
    ResourceHandlerReturn,
} from '@cloudcomponents/custom-resource-helper';

import { createWebhook, updateWebhook, deleteWebhook } from './webhook-api';

const getProperties = (event: CloudFormationCustomResourceEvent) => {
    const props = event.ResourceProperties;

    const githubApiToken = props.GithubApiToken;
    const githubRepoUrl = props.GithubRepoUrl;
    const payloadUrl = props.PayloadUrl;
    const events = props.Events;

    return {
        githubApiToken,
        githubRepoUrl,
        payloadUrl,
        events,
    };
};

const handleCreate: OnCreateHandler = async (
    event,
    _,
): Promise<ResourceHandlerReturn> => {
    const { githubApiToken, githubRepoUrl, payloadUrl, events } = getProperties(
        event,
    );
    const { data: responseData } = await createWebhook(
        githubApiToken,
        githubRepoUrl,
        payloadUrl,
        events,
    );

    const physicalResourceId = responseData.id.toString();

    return {
        physicalResourceId,
        responseData,
    };
};

const handleUpdate: OnUpdateHandler = async (
    event,
    _,
): Promise<ResourceHandlerReturn> => {
    const { githubApiToken, githubRepoUrl, payloadUrl, events } = getProperties(
        event,
    );
    const hookId = event.PhysicalResourceId;

    const { data: responseData } = await updateWebhook(
        githubApiToken,
        githubRepoUrl,
        payloadUrl,
        events,
        parseInt(hookId, 10),
    );

    const physicalResourceId = responseData.id.toString();

    return {
        physicalResourceId,
        responseData,
    };
};

const handleDelete: OnDeleteHandler = async (event, _): Promise<void> => {
    const { githubApiToken, githubRepoUrl } = getProperties(event);
    const hookId = event.PhysicalResourceId;

    await deleteWebhook(githubApiToken, githubRepoUrl, parseInt(hookId, 10));
};

export const handler = customResourceHelper(
    (): ResourceHandler => ({
        onCreate: handleCreate,
        onUpdate: handleUpdate,
        onDelete: handleDelete,
    }),
);
