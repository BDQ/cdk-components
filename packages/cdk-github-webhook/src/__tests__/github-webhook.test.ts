import { Stack } from '@aws-cdk/core';
import '@cloudcomponents/jest-cdk-snapshot';
import { GithubWebhook } from '../github-webhook';

describe('cdk-github-webhook', (): void => {
    it('snapshot', (): void => {
        const stack = new Stack();

        new GithubWebhook(stack, 'GithubWebhook', {
            githubApiToken: 'test12',
            githubRepoUrl: 'test',
            payloadUrl: 'test',
            events: ['test'],
        });

        expect(stack).toMatchCdkSnapshot();
    });
});
