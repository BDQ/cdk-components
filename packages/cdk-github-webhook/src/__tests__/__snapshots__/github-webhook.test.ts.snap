// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cdk-github-webhook snapshot 1`] = `
Object {
  "Parameters": Object {
    "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232ECodeArtifactHash7626A53A": Object {
      "Description": "Artifact hash for asset \\"Custom::GithubWebhook83CBF3EB7B6244F28C678441E4C1232E/Code\\"",
      "Type": "String",
    },
    "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232ECodeS3BucketB7B66645": Object {
      "Description": "S3 bucket for asset \\"Custom::GithubWebhook83CBF3EB7B6244F28C678441E4C1232E/Code\\"",
      "Type": "String",
    },
    "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232ECodeS3VersionKeyCF7474D7": Object {
      "Description": "S3 key for asset version \\"Custom::GithubWebhook83CBF3EB7B6244F28C678441E4C1232E/Code\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232EA4E17075": Object {
      "DependsOn": Array [
        "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232EServiceRoleA1A794DD",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232ECodeS3BucketB7B66645",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232ECodeS3VersionKeyCF7474D7",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232ECodeS3VersionKeyCF7474D7",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "lib/github-webhook.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232EServiceRoleA1A794DD",
            "Arn",
          ],
        },
        "Runtime": "nodejs10.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232EServiceRoleA1A794DD": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "lambda.",
                      Object {
                        "Ref": "AWS::URLSuffix",
                      },
                    ],
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "GithubWebhookCustomResourceA8270ACA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "Events": Array [
          "test",
        ],
        "GithubApiToken": "test12",
        "GithubRepoUrl": "test",
        "PayloadUrl": "test",
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomGithubWebhook83CBF3EB7B6244F28C678441E4C1232EA4E17075",
            "Arn",
          ],
        },
      },
      "Type": "Custom::GithubWebhook",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
